substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -487,35 +487,62 @@ where
 			"fatp::ready_at_light"
 		);
 
-		let Ok(block_number) = self.api.resolve_block_number(at) else {
+		// Return an empty ready transactions set if we can not find the number of `at` block hash.
+		let Ok(at_number) = self.api.resolve_block_number(at) else {
 			return Box::new(std::iter::empty())
 		};
 
-		let best_result = {
-			api.tree_route(self.enactment_state.lock().recent_finalized_block(), at).map(
-				|tree_route| {
-					if let Some((index, view)) =
-						tree_route.enacted().iter().enumerate().rev().skip(1).find_map(|(i, b)| {
-							self.view_store.get_view_at(b.hash, true).map(|(view, _)| (i, view))
-						}) {
-						let e = tree_route.enacted()[index..].to_vec();
-						(TreeRoute::new(e, 0).ok(), Some(view))
-					} else {
-						(None, None)
-					}
-				},
-			)
+		let mut to_process = Vec::new();
+		let mut best_view = None;
+		let mut tmp_at_number = at_number;
+		let mut tmp_at = at;
+
+		// Return an empty ready transaction set if we can't find a finalized block to report to,
+		// given we can not look for a view that can provide an approximate ready transaction set
+		// on the fork ending with `at` up to the finalized block.
+		let last_finalized = self.enactment_state.lock().recent_finalized_block();
+		let Ok(last_finalized_number) = self.api.resolve_block_number(last_finalized) else {
+			return Box::new(std::iter::empty())
 		};
 
-		if let Ok((Some(best_tree_route), Some(best_view))) = best_result {
+		// Search for a view that can be used to get and return an approximate ready
+		// transaction set.
+		while tmp_at_number > last_finalized_number {
+			// Found a view, stop searching..
+			if let Some((view, _)) = self.view_store.get_view_at(tmp_at, true) {
+				best_view = Some(view);
+				break;
+			}
+
+			to_process.push(tmp_at);
+			// Move up the hierarchy by getting the parent block.
+			if let Ok(tmp_at_header) = self.api.block_header(tmp_at) {
+				let Some(parent_number) = tmp_at_header.and_then(|header| {
+					tmp_at = *header.parent_hash();
+					self.api.resolve_block_number(*header.parent_hash()).ok()
+				}) else {
+					break;
+				};
+
+				tmp_at_number = parent_number;
+			} else {
+				// Not finding the header of any block on the fork means we can not continue looking
+				// for a view up to the finalized block, so we break early.
+				break;
+			}
```

Comment:
maybe sth like this will more readable - we have tmp_at_xxx assigned in one block and one break.
```suggestion
			let parent_resolved = self
				.api
				.block_header(tmp_at)
				.ok()
				.flatten()
				.and_then(|header| {
					let hash = *header.parent_hash();
					self.api.resolve_block_number(hash).ok().map(|number| (hash, number))
				})
				.map(|(parent_hash, parent_number)| {
					tmp_at = parent_hash;
					tmp_at_number = parent_number;
				});

			if parent_resolved.is_none() {
				break
			}

```

Reply:
Hmm, I actually went with something like below, which should work fine, and looks readable too.

```
   // Move up into the fork. Return with no view or enacted blocks list if
   // we can't access the header of the current block.
   let Some(header) = self.api.block_header(at_hash).ok().flatten() else {
      return (None, Vec::new());
   };
   at_hash = *header.parent_hash();
   at_number = at_number.saturating_sub(1u32.into());
```